### PR TITLE
unix,stream: clear read/write states on close

### PR DESCRIPTION
**Pull request by vtjnash**
Rebased copy of #1825 with some minor style cleanup

### CODE REVIEW

File: src/unix/stream.c [hunk 1]
```diff
@@ -1001,6 +1001,7 @@ uv_handle_type uv__handle_type(int fd) {
 static void uv__stream_eof(uv_stream_t* stream, const uv_buf_t* buf) {
   stream->flags |= UV_HANDLE_READ_EOF;
   stream->flags &= ~UV_HANDLE_READING;
+  stream->flags &= ~UV_HANDLE_READABLE;
```
Comment (saghul) [inline src/unix/stream.c]: how about writable?

File: test/test-not-readable-nor-writable-on-read-error.c [hunk 2]
```diff
@@ -0,0 +1,104 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Comment (saghul) [inline test/test-not-readable-nor-writable-on-read-error.c]: libuv contributors

File: test/echo-server.c [hunk 3]
```diff
@@ -68,6 +68,24 @@ static void after_shutdown(uv_shutdown_t* req, int status) {
   free(req);
 }
 
+static void dont_linger(uv_handle_t* handle) {
+  struct linger so_linger;
+  uv_os_fd_t socket_fd = 0;
```
Comment (saghul) [inline test/echo-server.c]: initialize to -1 for safety?

File: test/echo-server.c [hunk 4]
```diff
@@ -68,6 +68,24 @@ static void after_shutdown(uv_shutdown_t* req, int status) {
...
+
+  if (UV_TCP != uv_handle_get_type(handle)) {
+    return;
+  }
+
+  ASSERT(0 == uv_fileno(handle, &socket_fd));
+
+  so_linger.l_onoff = 1;
+  so_linger.l_linger = 0;
+  ASSERT(0 == setsockopt((uv_os_sock_t) socket_fd,
```
Comment (saghul) [inline test/echo-server.c]: no problems with this on Windows?

File: test/test-not-readable-nor-writable-on-read-error.c [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  connect_cb_called++;
+  ASSERT(status == 0);
+
+  r = uv_read_start((uv_stream_t*) &tcp_client, alloc_cb, read_cb);
+  ASSERT(r == 0);
+
+  reset_me = uv_buf_init(reset_me_cmd, sizeof(reset_me_cmd));
+
+  r = uv_write(&write_req,
+                      (uv_stream_t*) &tcp_client,
```
Comment (saghul) [inline test/test-not-readable-nor-writable-on-read-error.c]: style: align arguments

File: test/test-not-readable-on-eof.c [hunk 6]
```diff
@@ -0,0 +1,103 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Comment (saghul) [inline test/test-not-readable-on-eof.c]: libuv contributors

File: test/test-not-readable-on-eof.c [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+static void close_cb(uv_handle_t* handle) {
+  close_cb_called++;
+}
+
+static void read_cb(uv_stream_t* handle, ssize_t nread, const uv_buf_t* buf) {
+  read_cb_called++;
+
+  ASSERT(nread == UV_EOF);
+  ASSERT(0 == uv_is_readable(handle));
```
Comment (saghul) [inline test/test-not-readable-on-eof.c]: check also if it's writable?

File: test/test-not-readable-on-eof.c [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  connect_cb_called++;
+  ASSERT(status == 0);
+
+  r = uv_read_start((uv_stream_t*) &tcp_client, alloc_cb, read_cb);
+  ASSERT(r == 0);
+
+  close_me = uv_buf_init(close_me_cmd, sizeof(close_me_cmd));
+
+  r = uv_write(&write_req,
+                      (uv_stream_t*) &tcp_client,
```
Comment (saghul) [inline test/test-not-readable-on-eof.c]: style: align args

File: test/test-not-writable-after-shutdown.c [hunk 9]
```diff
@@ -0,0 +1,69 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Comment (saghul) [inline test/test-not-writable-after-shutdown.c]: ditto

File: test/test.gyp [hunk 10]
```diff
@@ -162,6 +162,9 @@
         'test-udp-multicast-interface6.c',
         'test-udp-try-send.c',
         'test-uname.c',
+        'test-not-writable-after-shutdown.c',
```
Comment (saghul) [inline test/test.gyp]: sort alphabetically

File: test/test.gyp [hunk 11]
```diff
@@ -162,6 +162,9 @@
...
+        'test-not-readable-nor-writable-on-read-error.c',
+        'test-not-readable-on-eof.c'
```
Comment (saghul) [inline test/test.gyp]: we need these in the CMakeLists too

File: test/test-not-readable-nor-writable-on-read-error.c [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+ */
+
+#include "uv.h"
+#include "task.h"
+
+uv_loop_t loop;
+uv_tcp_t tcp_client;
+uv_connect_t connect_req;
+uv_write_t write_req;
+char reset_me_cmd[3] = {'Q', 'S', 'H'};
```
Comment (bnoordhuis) [inline test/test-not-readable-nor-writable-on-read-error.c]: These should all be `static`.

File: test/test-not-readable-on-eof.c [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+ */
+
+#include "uv.h"
+#include "task.h"
+
+uv_loop_t loop;
+uv_tcp_t tcp_client;
+uv_connect_t connect_req;
+uv_write_t write_req;
+char close_me_cmd[2] = {'Q', 'S'};
```
Comment (bnoordhuis) [inline test/test-not-readable-on-eof.c]: Ditto.

File: src/unix/stream.c (hunk 1 above)
Comment (vtjnash) [reply src/unix/stream.c]: Does reaching the end of the input make a stream immediately non-writable? That feels false (e.g. what about `shutdown()`), although I often do end up coupling them too.

File: test/test-not-readable-nor-writable-on-read-error.c (hunk 2 above)
Comment (vtjnash) [reply test/test-not-readable-nor-writable-on-read-error.c]: `Copyright the libuv project contributors. All rights reserved.` ? That's used on 5 other files. All of the rest have this header.

File: test/test.gyp (hunk 10 above)
Comment (vtjnash) [reply test/test.gyp]: Is is worthwhile to sort the whole list or to make a separate PR for doing that?

File: test/echo-server.c (hunk 4 above)
Comment (vtjnash) [reply test/echo-server.c]: This capability is now equivalent to `uv_tcp_close_reset`, so I suppose it should be fine. I've refactored it to use `uv_tcp_close_reset` directly now.

**Comment by vtjnash** [comment]
https://ci.nodejs.org/job/libuv-test-commit/1568/

File: test/test-not-readable-nor-writable-on-read-error.c (hunk 2 above)
Comment (bnoordhuis) [reply test/test-not-readable-nor-writable-on-read-error.c]: Yes, that's the one. New files get the new copyright boilerplate.

File: test/test.gyp [hunk 14]
```diff
@@ -10,85 +10,92 @@
...
+        'test-loop-handles.c',
         'test-loop-stop.c',
         'test-loop-time.c',
-        'test-loop-configure.c',
-        'test-walk-handles.c',
-        'test-watcher-cross-stop.c',
         'test-multiple-listen.c',
+        'test-mutexes.c',
+        'test-not-readable-nor-writable-on-read-error.c',
+        'test-not-readable-on-eof.c'
```
Comment (bnoordhuis) [inline test/test.gyp]: An interesting factoid about gyp files: if you omit the comma, it concatenates the strings. ^^

(The build failures are caused by the missing comma here.)

**Comment by vtjnash** [comment]
CI: https://ci.nodejs.org/job/libuv-test-commit/1673/

@saghul did you want to re-review? I'm assuming I should just go ahead and land this, since I think I addressed all comments (and @bnoordhuis subsequently approved).

**Comment by vtjnash** [comment]
It appears that the new `uv_tcp_close_reset` function may not work on AIX:

```
not ok 161 - not_readable_nor_writable_on_read_error
# exit code 393222
# Output from process `not_readable_nor_writable_on_read_error`: (no output)
Output from process `not_readable_nor_writable_on_read_error`:
# Assertion failed in test/test-not-readable-nor-writable-on-read-error.c on line 56: (nread < 0) && (nread != UV_EOF)
```

**Comment by vtjnash** [comment]
Ah, digging further, I see that AIX is now (un)intentionally supposed to fail to handle this error condition (https://github.com/libuv/libuv/pull/2447). If I revert that PR, AIX starts emitting errors correctly again (8ab6a79 / https://ci.nodejs.org/job/libuv-test-commit-aix/1717/). I'll disable this test on AIX and open an issue.

**Comment by vtjnash** [comment]
new CI: https://ci.nodejs.org/job/libuv-test-commit/1679/

**Comment by vtjnash** [comment]
Node CI: https://ci.nodejs.org/view/libuv/job/libuv-in-node/

**Comment by vtjnash** [comment]
Can someone with knowledge of nodejs tests help me interpret that result? I see several bots failed running this test:
```
14:05:03 not ok 2529 parallel/test-http2-stream-destroy-event-order
14:05:03   ---
14:05:03   duration_ms: 120.35
14:05:03   severity: fail
14:05:03   exitcode: -15
14:05:03   stack: |-
14:05:03     timeout
14:05:03   ...
```
https://github.com/nodejs/node/blob/f5512ff61ecb668c2f49b7c05d3227ef7aa5e85f/test/parallel/test-http2-stream-destroy-event-order.js

But maybe that's a known problem? https://github.com/nodejs/node/pull/29889#issuecomment-543726136

**Comment by vtjnash** [comment]
Bump???

Also it maybe also hit a more serious (unrelated) bug in nodejs? (https://ci.nodejs.org/job/node-test-commit-arm/28686/nodes=debian9-docker-armv7/)
```
not ok 2700 sequential/test-http2-ping-flood
  ---
  duration_ms: 3.457
  severity: crashed
  exitcode: -4
  stack: |-
  ...
```

**Comment by cjihrig** [comment]
The failures are at least not completely related to your changes here.

test-http2-stream-destroy-event-order is being tracked as a flaky test in Node in https://github.com/nodejs/node/issues/31281.

test-http2-ping-flood is being tracked as a flaky test in Node in https://github.com/nodejs/node/issues/31107.

**Comment by vtjnash** [comment]
OK, good to know those are probably unrelated (or at least, pre-existing). I think it should be good to land now (esp. now that 1.34.1 is branched).

Important note though: the merging-person should be aware there's a potential merge conflict with #2602: this adds a test checking for AIX being broken in the way that #2602 would fix. Thus whichever one lands first will imply a need to update and re-CI the other one.

**Comment by vtjnash** [comment]
libuv CI: https://ci.nodejs.org/job/libuv-test-commit/1722/ ✅ 
node CI: https://ci.nodejs.org/view/libuv/job/libuv-in-node/123/ ⚠️

**Comment by cjihrig** [comment]
FYI - this was reverted in https://github.com/libuv/libuv/pull/2968.

**Pull request merged**